pkg/logging/log_viewer.go

```diff
@@ -0,0 +1,145 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package logging
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/sirupsen/logrus"
+)
+
+// Type alias for functions which write out logs to the provided stdout/stderr Writers.
+// Depending on the provided `LogViewOptions.Follow` option, the function may block
+// indefinitely until something is sent through the `stopChannel`.
+type LogViewerFunc func(lvopts LogViewOptions, stdout, stderr io.Writer, stopChannel chan os.Signal) error
+
+var logViewers = make(map[string]LogViewerFunc)
+
+// Registers a LogViewerFunc for the
+func RegisterLogViewer(driverName string, lvfn LogViewerFunc) {
+	if v, ok := logViewers[driverName]; ok {
+		logrus.Warnf("A LogViewerFunc with name %q has already been registered: %#v, overriding with %#v either way", driverName, v, lvfn)
+	}
+	logViewers[driverName] = lvfn
+}
+
+func init() {
+	RegisterLogViewer("json-file", viewLogsJSONFile)
+	RegisterLogViewer("journald", viewLogsJournald)
```

Comment:
what about `fluentd` still supported ?

Reply:
As of the making of this patch, it [doesn't look like viewing logs from `fluentd` is supported](https://github.com/containerd/nerdctl/blob/main/cmd/nerdctl/logs.go#L120-L150) in `nerdctl logs`.
Specifying `nerdctl run --log-driver fluend` seems to be supported, as [it's even tested](https://github.com/containerd/nerdctl/blob/main/cmd/nerdctl/run_linux_test.go#L202-L234), but [the tests read the logs from the fluentd files directly](https://github.com/containerd/nerdctl/blob/main/cmd/nerdctl/run_linux_test.go#L255-L262) instead of validating the output of `nerdctl logs` like the other logging-related tests.

It's probably a bit late for me to mention this, but my immediate goal (before I saw [this TODO](https://github.com/containerd/nerdctl/blob/main/cmd/nerdctl/logs.go#L119)) was getting containerd logging and `nerdctl logs` to work on Windows, so adding log viewing for fluentd/syslog would be even more beyond my initial scope 😆 

I had considered the need for adding `fluentd` and `syslog` log viewers down the line which is why I tried to make the logging viewers as pluggable and extensible as possible. (hypothetically all the logging options needed to read the logs for a given container [will be saved in the logging config on container creation](https://github.com/containerd/nerdctl/blob/main/cmd/nerdctl/run.go#L527) and loaded in [`LogViewer.loggingConfig`](https://github.com/containerd/nerdctl/pull/1468/files#diff-eabf36f45659163d25efc51651ed872d0d7d96de5fe03318b01dcec12cc1c726R118-R123) so the LogViewer can go from there...)

Reply:
SGTM